File: backend/app/schemas/monitoring.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from db.monitoring_models import (
    ALERT_SEVERITIES,
    ALERT_STATUSES,
    CHANNEL_TYPES,
    FREQUENCIES,
)


# ── Channels ────────────────────────────────────────────────────────────────

class ChannelTypeInfo(BaseModel):
    type: str
    label: str
    description: str
    destination_label: str = ""
    destination_placeholder: str = ""
    needs_destination: bool = True
    secret_fields: list[dict[str, Any]] = Field(default_factory=list)
    config_fields: list[dict[str, Any]] = Field(default_factory=list)


class ChannelOut(BaseModel):
    """A saved channel. Credentials are never echoed back — only whether some
    are stored, which is all the UI needs to render "configured"."""

    id: str
    type: str
    name: str
    destination: str
    config: dict[str, Any] = Field(default_factory=dict)
    is_active: bool
    has_credentials: bool
    verified_at: datetime | None = None
    last_error: str | None = None
    created_at: datetime


class CreateChannelRequest(BaseModel):
    type: str
    name: str
    destination: str = ""
    credentials: dict[str, str] = Field(default_factory=dict)
    config: dict[str, Any] = Field(default_factory=dict)
    is_active: bool = True

    @field_validator("type")
    @classmethod
    def _known_type(cls, value: str) -> str:
        if value not in CHANNEL_TYPES:
            raise ValueError(f"Unknown channel type '{value}'. One of: {', '.join(CHANNEL_TYPES)}")
        return value


class UpdateChannelRequest(BaseModel):
    name: str | None = None
    destination: str | None = None
    # Omitted leaves the stored credentials untouched; sending {} clears them.
    credentials: dict[str, str] | None = None
    config: dict[str, Any] | None = None
    is_active: bool | None = None


class ChannelTestResponse(BaseModel):
    ok: bool
    channel_type: str
    destination: str = ""
    error: str | None = None
    provider_message_id: str | None = None


# ── Schedules ───────────────────────────────────────────────────────────────

class ScheduleOut(BaseModel):
    id: str
    project_id: str
    project_name: str
    name: str
    is_active: bool
    frequency: str
    hour: int
    minute: int
    day_of_week: int
    day_of_month: int
    cron_expression: str | None = None
    timezone: str
    analyses: list[str] = Field(default_factory=list)
    rules: dict[str, Any] = Field(default_factory=dict)
    min_severity: str
    cooldown_hours: int
    max_alerts_per_run: int
    channel_ids: list[str] = Field(default_factory=list)
    language: str
    quiet_hours_start: int | None = None
    quiet_hours_end: int | None = None
    send_when_nothing_found: bool
    last_run_at: datetime | None = None
    last_status: str | None = None
    next_run_at: datetime | None = None
    created_at: datetime
    # Rendered by the backend so the UI never has to reimplement cron semantics
    # — and so what the user reads back is what will actually fire.
    description: str = ""


class CreateScheduleRequest(BaseModel):
    name: str = "Daily check"
    is_active: bool = True
    frequency: str = "daily"
    hour: int = Field(default=7, ge=0, le=23)
    minute: int = Field(default=0, ge=0, le=59)
    day_of_week: int = Field(default=0, ge=0, le=6)
    day_of_month: int = Field(default=1, ge=1, le=28)
    cron_expression: str | None = None
    timezone: str = "UTC"
    analyses: list[str] = Field(default_factory=lambda: ["insights", "root_cause"])
    rules: dict[str, Any] = Field(default_factory=dict)
    min_severity: str = "medium"
    cooldown_hours: int = Field(default=24, ge=0, le=720)
    max_alerts_per_run: int = Field(default=5, ge=1, le=25)
    channel_ids: list[str] = Field(default_factory=list)
    language: str = "en"
    quiet_hours_start: int | None = Field(default=None, ge=0, le=23)
    quiet_hours_end: int | None = Field(default=None, ge=0, le=23)
    send_when_nothing_found: bool = False

    @field_validator("frequency")
    @classmethod
    def _known_frequency(cls, value: str) -> str:
        if value not in FREQUENCIES:
            raise ValueError(f"Unknown frequency '{value}'. One of: {', '.join(FREQUENCIES)}")
        return value

    @field_validator("min_severity")
    @classmethod
    def _known_severity(cls, value: str) -> str:
        if value not in ALERT_SEVERITIES:
            raise ValueError(
                f"Unknown severity '{value}'. One of: {', '.join(ALERT_SEVERITIES)}"
            )
        return value


class UpdateScheduleRequest(BaseModel):
    name: str | None = None
    is_active: bool | None = None
    frequency: str | None = None
    hour: int | None = Field(default=None, ge=0, le=23)
    minute: int | None = Field(default=None, ge=0, le=59)
    day_of_week: int | None = Field(default=None, ge=0, le=6)
    day_of_month: int | None = Field(default=None, ge=1, le=28)
    cron_expression: str | None = None
    timezone: str | None = None
    analyses: list[str] | None = None
    rules: dict[str, Any] | None = None
    min_severity: str | None = None
    cooldown_hours: int | None = Field(default=None, ge=0, le=720)
    max_alerts_per_run: int | None = Field(default=None, ge=1, le=25)
    channel_ids: list[str] | None = None
    language: str | None = None
    quiet_hours_start: int | None = Field(default=None, ge=0, le=23)
    quiet_hours_end: int | None = Field(default=None, ge=0, le=23)
    send_when_nothing_found: bool | None = None


# ── Runs and alerts ─────────────────────────────────────────────────────────

class RunOut(BaseModel):
    id: str
    schedule_id: str
    schedule_name: str = ""
    project_id: str
    project_name: str = ""
    status: str
    trigger: str
    started_at: datetime
    finished_at: datetime | None = None
    duration_ms: int | None = None
    alerts_found: int
    alerts_delivered: int
    money_at_stake: float
    error: str | None = None


class RunDetailOut(RunOut):
    briefing_md: str = ""
    detail: dict[str, Any] = Field(default_factory=dict)
    alerts: list[AlertOut] = Field(default_factory=list)
    deliveries: list[dict[str, Any]] = Field(default_factory=list)


class AlertOut(BaseModel):
    id: str
    run_id: str
    project_id: str
    project_name: str = ""
    fingerprint: str
    rule: str
    severity: str
    title: str
    body_md: str
    metric: str
    money_at_stake: float
    current_value: float | None = None
    prior_value: float | None = None
    change_pct: float | None = None
    evidence: dict[str, Any] = Field(default_factory=dict)
    root_cause: dict[str, Any] = Field(default_factory=dict)
    status: str
    read_at: datetime | None = None
    created_at: datetime


class UpdateAlertRequest(BaseModel):
    status: str

    @field_validator("status")
    @classmethod
    def _known_status(cls, value: str) -> str:
        if value not in ALERT_STATUSES:
            raise ValueError(f"Unknown status '{value}'. One of: {', '.join(ALERT_STATUSES)}")
        return value


class AlertSummaryResponse(BaseModel):
    unread: int
    recent: list[AlertOut] = Field(default_factory=list)


class PreviewResponse(BaseModel):
    """What a schedule *would* have sent, without sending or recording it."""

    available: bool
    reason: str = ""
    subject: str = ""
    briefing_md: str = ""
    text: str = ""
    money_at_stake: float = 0.0
    findings: list[dict[str, Any]] = Field(default_factory=list)
    suppressed: list[dict[str, Any]] = Field(default_factory=list)
    root_cause: dict[str, Any] = Field(default_factory=dict)
    would_deliver: bool = False
    channels: list[dict[str, Any]] = Field(default_factory=list)


class SchedulerStatusResponse(BaseModel):
    running: bool
    reason: str = ""
    jobs: list[dict[str, Any]] = Field(default_factory=list)


RunDetailOut.model_rebuild()

File: frontend/src/lib/queries/monitoring.ts
'use client'

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { useApi } from '@/lib/api/use-api'

export type Severity = 'critical' | 'high' | 'medium' | 'low' | 'info'
export type AlertStatus = 'new' | 'acknowledged' | 'resolved' | 'muted'

export type ChannelField = {
  key: string
  label: string
  placeholder?: string
  options?: string[]
  /** Only shown when the channel's `provider` config matches. */
  provider?: string
}

export type ChannelTypeInfo = {
  type: string
  label: string
  description: string
  destination_label: string
  destination_placeholder: string
  needs_destination: boolean
  secret_fields: ChannelField[]
  config_fields: ChannelField[]
}

export type Channel = {
  id: string
  type: string
  name: string
  destination: string
  config: Record<string, any>
  is_active: boolean
  /** Credentials are never returned — only whether some are stored. */
  has_credentials: boolean
  verified_at: string | null
  last_error: string | null
  created_at: string
}

export type Schedule = {
  id: string
  project_id: string
  project_name: string
  name: string
  is_active: boolean
  frequency: string
  hour: number
  minute: number
  day_of_week: number
  day_of_month: number
  cron_expression: string | null
  timezone: string
  analyses: string[]
  rules: Record<string, any>
  min_severity: Severity
  cooldown_hours: number
  max_alerts_per_run: number
  channel_ids: string[]
  language: string
  quiet_hours_start: number | null
  quiet_hours_end: number | null
  send_when_nothing_found: boolean
  last_run_at: string | null
  last_status: string | null
  next_run_at: string | null
  created_at: string
  /** Rendered server-side, so the UI never reimplements cron semantics. */
  description: string
}

export type MonitorRun = {
  id: string
  schedule_id: string
  schedule_name: string
  project_id: string
  project_name: string
  status: 'pending' | 'running' | 'success' | 'failed' | 'skipped'
  trigger: 'schedule' | 'manual'
  started_at: string
  finished_at: string | null
  duration_ms: number | null
  alerts_found: number
  alerts_delivered: number
  money_at_stake: number
  error: string | null
}

export type Alert = {
  id: string
  run_id: string
  project_id: string
  project_name: string
  fingerprint: string
  rule: string
  severity: Severity
  title: string
  body_md: string
  metric: string
  money_at_stake: number
  current_value: number | null
  prior_value: number | null
  change_pct: number | null
  evidence: Record<string, any>
  root_cause: Record<string, any>
  status: AlertStatus
  read_at: string | null
  created_at: string
}

export type RunDetail = MonitorRun & {
  briefing_md: string
  detail: Record<string, any>
  alerts: Alert[]
  deliveries: {
    id: string
    channel_type: string
    status: string
    destination: string
    error: string | null
    attempts: number
    created_at: string | null
  }[]
}

export type Preview = {
  available: boolean
  reason: string
  subject: string
  briefing_md: string
  text: string
  money_at_stake: number
  findings: {
    rule: string
    severity: Severity
    title: string
    body: string
    money_at_stake: number
    metric: string
  }[]
  suppressed: { rule: string; severity: Severity; title: string }[]
  root_cause: Record<string, any>
  would_deliver: boolean
  channels: { id: string; type: string; name: string; active: boolean }[]
}

export type SchedulerStatus = {
  running: boolean
  reason: string
  jobs: { id: string; name: string; next_run_at: string | null }[]
}

export type AlertSummary = {
  unread: number
  recent: Alert[]
}

// ── Channels ───────────────────────────────────────────────────────────────

export function useChannelTypes() {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['monitoring', 'channel-types'],
    queryFn: () => call<ChannelTypeInfo[]>('/api/v1/monitoring/channel-types'),
    enabled: ready,
    staleTime: Infinity,
  })
}

export function useChannels() {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['monitoring', 'channels'],
    queryFn: () => call<Channel[]>('/api/v1/monitoring/channels'),
    enabled: ready,
  })
}

export function useCreateChannel() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (input: {
      type: string
      name: string
      destination?: string
      credentials?: Record<string, string>
      config?: Record<string, any>
      is_active?: boolean
    }) => call<Channel>('/api/v1/monitoring/channels', { method: 'POST', json: input }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring', 'channels'] }),
  })
}

export function useUpdateChannel() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...input }: { id: string } & Record<string, any>) =>
      call<Channel>(`/api/v1/monitoring/channels/${id}`, { method: 'PATCH', json: input }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring', 'channels'] }),
  })
}

export function useDeleteChannel() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      call<void>(`/api/v1/monitoring/channels/${id}`, { method: 'DELETE' }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring', 'channels'] }),
  })
}

export function useTestChannel() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      call<{ ok: boolean; channel_type: string; destination: string; error: string | null }>(
        `/api/v1/monitoring/channels/${id}/test`,
        { method: 'POST' }
      ),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring', 'channels'] }),
  })
}

// ── Schedules ──────────────────────────────────────────────────────────────

export function useSchedules() {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['monitoring', 'schedules'],
    queryFn: () => call<Schedule[]>('/api/v1/monitoring/schedules'),
    enabled: ready,
  })
}

export function useCreateSchedule(projectId: string) {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (input: Record<string, any>) =>
      call<Schedule>(`/api/v1/monitoring/projects/${projectId}/schedules`, {
        method: 'POST',
        json: input,
      }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring'] }),
  })
}

export function useUpdateSchedule() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...input }: { id: string } & Record<string, any>) =>
      call<Schedule>(`/api/v1/monitoring/schedules/${id}`, { method: 'PATCH', json: input }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring'] }),
  })
}

export function useDeleteSchedule() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      call<void>(`/api/v1/monitoring/schedules/${id}`, { method: 'DELETE' }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring'] }),
  })
}

export function usePreviewSchedule() {
  const { call } = useApi()
  return useMutation({
    mutationFn: (id: string) =>
      call<Preview>(`/api/v1/monitoring/schedules/${id}/preview`, { method: 'POST' }),
  })
}

export function useRunScheduleNow() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      call<MonitorRun>(`/api/v1/monitoring/schedules/${id}/run`, { method: 'POST' }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring'] }),
  })
}

// ── Runs and alerts ────────────────────────────────────────────────────────

export function useRuns(limit = 50) {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['monitoring', 'runs', limit],
    queryFn: () => call<MonitorRun[]>(`/api/v1/monitoring/runs?limit=${limit}`),
    enabled: ready,
  })
}

export function useRun(runId: string | null) {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['monitoring', 'run', runId],
    queryFn: () => call<RunDetail>(`/api/v1/monitoring/runs/${runId}`),
    enabled: ready && Boolean(runId),
  })
}

export function useAlerts(status?: AlertStatus) {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['monitoring', 'alerts', status ?? 'all'],
    queryFn: () =>
      call<Alert[]>(
        `/api/v1/monitoring/alerts${status ? `?status_filter=${status}` : ''}`
      ),
    enabled: ready,
  })
}

/** Drives the header bell. Polled, because alerts arrive from a scheduler the
 * browser has no other way of hearing from. */
export function useAlertSummary() {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['monitoring', 'alert-summary'],
    queryFn: () => call<AlertSummary>('/api/v1/monitoring/alerts/summary'),
    enabled: ready,
    refetchInterval: 60_000,
    refetchOnWindowFocus: true,
  })
}

export function useUpdateAlert() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, status }: { id: string; status: AlertStatus }) =>
      call<Alert>(`/api/v1/monitoring/alerts/${id}`, { method: 'PATCH', json: { status } }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring'] }),
  })
}

export function useMarkAllAlertsRead() {
  const { call } = useApi()
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () =>
      call<AlertSummary>('/api/v1/monitoring/alerts/read-all', { method: 'POST' }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['monitoring'] }),
  })
}

export function useSchedulerStatus() {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['monitoring', 'scheduler'],
    queryFn: () => call<SchedulerStatus>('/api/v1/monitoring/scheduler'),
    enabled: ready,
    refetchInterval: 120_000,
  })
}
